nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -422,97 +899,168 @@ impl DataStore {
+        assert_matches!(error, Error::ObjectNotFound { .. });
 
         let silo_id = opctx.authn.silo_required().unwrap().id();
 
         // create a non-default pool for the silo
         let identity = IdentityMetadataCreateParams {
-            name: "non-default-for-silo".parse().unwrap(),
+            name: "pool1-for-silo".parse().unwrap(),
             description: "".to_string(),
         };
+        let pool1_for_silo = datastore
+            .ip_pool_create(&opctx, IpPool::new(&identity))
+            .await
+            .expect("Failed to create IP pool");
         datastore
-            .ip_pool_create(
+            .ip_pool_link_silo(
                 &opctx,
-                IpPool::new(&identity, Some(silo_id), /*default= */ false),
+                IpPoolResource {
+                    ip_pool_id: pool1_for_silo.id(),
+                    resource_type: IpPoolResourceType::Silo,
+                    resource_id: silo_id,
+                    is_default: false,
+                },
             )
             .await
-            .expect("Failed to create silo non-default IP pool");
+            .expect("Failed to associate IP pool with silo");
 
         // because that one was not a default, when we ask for the silo default
-        // pool, we still get the fleet default
-        let ip_pool = datastore
-            .ip_pools_fetch_default(&opctx)
-            .await
-            .expect("Failed to get silo default IP pool");
-        assert_eq!(ip_pool.id(), fleet_default_pool.id());
+        // pool, we still get nothing
+        let error = datastore.ip_pools_fetch_default(&opctx).await.unwrap_err();
+        assert_matches!(error, Error::ObjectNotFound { .. });
 
-        // now create a default pool for the silo
-        let identity = IdentityMetadataCreateParams {
-            name: "default-for-silo".parse().unwrap(),
-            description: "".to_string(),
-        };
+        // now we can change that association to is_default=true and
+        // it should update rather than erroring out
         datastore
-            .ip_pool_create(&opctx, IpPool::new(&identity, Some(silo_id), true))
+            .ip_pool_link_silo(
+                &opctx,
+                IpPoolResource {
+                    ip_pool_id: pool1_for_silo.id(),
+                    resource_type: IpPoolResourceType::Silo,
+                    resource_id: silo_id,
+                    is_default: true,
+                },
+            )
             .await
-            .expect("Failed to create silo default IP pool");
+            .expect("Failed to make IP pool default for silo");
 
-        // now when we ask for the default pool, we get the one we just made
-        let ip_pool = datastore
+        // now when we ask for the default pool again, we get that one
+        let (authz_pool1_for_silo, ip_pool) = datastore
             .ip_pools_fetch_default(&opctx)
             .await
             .expect("Failed to get silo's default IP pool");
```

Comment:
I see, because this doesn't have a normal lookup implemented this fetch implementation acts as one. Seems fine.